src/main/java/com/amazonaws/encryptionsdk/kms/KmsMasterKeyProvider.java

```diff
@@ -32,76 +38,348 @@
 import com.amazonaws.encryptionsdk.exception.AwsCryptoException;
 import com.amazonaws.encryptionsdk.exception.NoSuchMasterKeyException;
 import com.amazonaws.encryptionsdk.exception.UnsupportedProviderException;
-import com.amazonaws.encryptionsdk.internal.VersionInfo;
-import com.amazonaws.internal.StaticCredentialsProvider;
+import com.amazonaws.handlers.RequestHandler2;
+import com.amazonaws.regions.DefaultAwsRegionProviderChain;
 import com.amazonaws.regions.Region;
 import com.amazonaws.regions.Regions;
 import com.amazonaws.services.kms.AWSKMS;
 import com.amazonaws.services.kms.AWSKMSClient;
+import com.amazonaws.services.kms.AWSKMSClientBuilder;
 
 /**
  * Provides {@link MasterKey}s backed by the AWS Key Management Service. This object is regional and
  * if you want to use keys from multiple regions, you'll need multiple copies of this object.
  */
 public class KmsMasterKeyProvider extends MasterKeyProvider<KmsMasterKey> implements KmsMethods {
-	private static final ClientConfiguration DEFAULT_CLIENT_CONFIG =
-			new ClientConfiguration().withUserAgentSuffix(VersionInfo.USER_AGENT);
     private static final String PROVIDER_NAME = "aws-kms";
-    private final AWSKMS kms_;
     private final List<String> keyIds_;
-    private final List<String> grantTokens_ = new ArrayList<>();
-    private Region region_;
-    private String regionName_;
+    private final List<String> grantTokens_;
+
+    private final RegionalClientSupplier regionalClientSupplier_;
+    private final String defaultRegion_;
+
+    @FunctionalInterface
+    public interface RegionalClientSupplier {
+        /**
+         * Supplies an AWSKMS instance to use for a given region. The {@link KmsMasterKeyProvider} will not cache the
+         * result of this function.
+         *
+         * @param regionName The region to get a client for
+         * @return The client to use, or null if this region cannot or should not be used.
+         */
+        AWSKMS getClient(String regionName);
+    }
+
+    public static class Builder implements Cloneable {
+        private String defaultRegion_ = null;
+        private RegionalClientSupplier regionalClientSupplier_ = null;
+        private AWSKMSClientBuilder templateBuilder_ = null;
```

Comment:
This should be copied when .clone() is invoked. A shallow copy here will likely cause problems.